### PR TITLE
[Label] Handle text as multiple independent paragraphs.

### PR DESCRIPTION
**Pull request by bruvzg**
See https://github.com/godotengine/godot/pull/79599

### CODE REVIEW

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected.

https://github.com/godotengine/godot/pull/79599 now performs much better when you replace the RichTextLabel with a Label and Word autowrap (note that I didn't set up the ScrollContainer). On an i9-13900K, it takes roughly 0.5s to open the licenses dialog with a debug editor build, whereas it took 3.5s previously.

RichTextLabel without threading is still noticeably faster though, even when compared to Label with autowrap disabled. With threading, there is no noticeable pause at all when opening the licenses dialog.

File: scene/gui/label.cpp [hunk 1]
```diff
@@ -747,24 +902,33 @@ int Label::get_line_count() const {
...
-	if (lines_visible > lines_rid.size()) {
-		lines_visible = lines_rid.size();
+	int line_index = 0;
+	for (const Paragraph &para : paragraphs) {
+		if (line_index + para.lines_rid.size() <= lines_skipped) {
+			line_index += para.lines_rid.size();
+		} else {
+			int start = (line_index < lines_skipped) ? lines_skipped - line_index : 0;
+			for (int i = start; i < para.lines_rid.size(); i++) {
+				total_h += TS->shaped_text_get_size(para.lines_rid[i]).y + line_spacing;
```
Comment (AThousandShips) [inline scene/gui/label.cpp:923]:
```suggestion
			for (const RID &line_rid : para.lines_rid) {
				total_h += TS->shaped_text_get_size(line_rid).y + line_spacing;
```

File: scene/gui/label.cpp [hunk 2]
```diff
@@ -1143,17 +1340,18 @@ void Label::_bind_methods() {
...
 	set_text(p_text);
 	set_v_size_flags(SIZE_SHRINK_CENTER);
 }
 
 Label::~Label() {
-	for (int i = 0; i < lines_rid.size(); i++) {
-		TS->free_rid(lines_rid[i]);
+	for (Paragraph &para : paragraphs) {
+		for (int i = 0; i < para.lines_rid.size(); i++) {
+			TS->free_rid(para.lines_rid[i]);
```
Comment (AThousandShips) [inline scene/gui/label.cpp:1351]:
```suggestion
		for (const RID &line_rid : para.lines_rid) {
			TS->free_rid(line_rid);
```

File: scene/gui/label.cpp [hunk 3]
```diff
@@ -105,181 +111,216 @@ void Label::_shape() {
...
+	int visible_lines = get_visible_line_count();
+	bool lines_hidden = visible_lines > 0 && visible_lines < total_line_count;
+
+	int line_index = 0;
+	for (Paragraph &para : paragraphs) {
+		if (autowrap_mode == TextServer::AUTOWRAP_OFF) {
+			minsize.width = 0.0f;
+			for (int i = 0; i < para.lines_rid.size(); i++) {
+				if (minsize.width < TS->shaped_text_get_size(para.lines_rid[i]).x) {
+					minsize.width = TS->shaped_text_get_size(para.lines_rid[i]).x;
```
Comment (AThousandShips) [inline scene/gui/label.cpp:223]:
```suggestion
			for (const RID &line_rid : para.lines_rid) {
				if (minsize.width < TS->shaped_text_get_size(line_rid).x) {
					minsize.width = TS->shaped_text_get_size(line_rid).x;
```

File: scene/gui/label.cpp [hunk 4]
```diff
@@ -381,73 +475,95 @@ Rect2 Label::get_line_rect(int p_line) const {
...
-		if (total_h > (get_size().height - style->get_minimum_size().height + line_spacing)) {
-			break;
+	int line_index = 0;
+	for (const Paragraph &para : paragraphs) {
+		if (line_index + para.lines_rid.size() <= lines_skipped) {
+			line_index += para.lines_rid.size();
+		} else {
+			int start = (line_index < lines_skipped) ? lines_skipped - line_index : 0;
+			for (int i = start; i < para.lines_rid.size(); i++) {
+				total_h += TS->shaped_text_get_size(para.lines_rid[i]).y + line_spacing;
```
Comment (AThousandShips) [inline scene/gui/label.cpp:500]:
```suggestion
			for (const RID &line_rid : para.lines_rid) {
				total_h += TS->shaped_text_get_size(line_rid).y + line_spacing;
```

**Review by AThousandShips — APPROVED** [review]

File: scene/gui/label.cpp (hunk 4 above)
Comment (bruvzg) [reply scene/gui/label.cpp:500]: It's not iterating over a full range here.

File: scene/gui/label.cpp (hunk 1 above)
Comment (bruvzg) [reply scene/gui/label.cpp:923]: Same.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**